akka-remote/src/main/scala/akka/remote/artery/tcp/ArteryTcpTransport.scala

```diff
@@ -326,14 +326,15 @@ private[remote] class ArteryTcpTransport(
     // overhead.
     val inboundStream = Sink.fromGraph(GraphDSL.create() { implicit b =>
       import GraphDSL.Implicits._
-      val partition = b.add(Partition[EnvelopeBuffer](3, env => {
+      val partitioner: EnvelopeBuffer => Int = env => {
         env.streamId match {
           case OrdinaryStreamId => 1
           case ControlStreamId  => 0
           case LargeStreamId    => 2
           case other            => throw new IllegalArgumentException(s"Unexpected streamId [$other]")
         }
-      }))
+      }
+      val partition = b.add(Partition[EnvelopeBuffer](3,  partitioner = partitioner , eagerCancel = false))
```

Comment:
If this was required to not fail the build it means the improvement/apply overload broke source compatibility and could break user code as well, which is not ok.

Reply:
yeah, that is true. Any idea why it only breaks in Scala 2.12? Further,  I know that Alpakka is dropping support for 2.12 in the next major release. Is the same true for Akka, and maybe this PR could be interesting then?

Reply:
Not sure exactly why this happens, perhaps you could try moving the eager parameter in the overload so the lambda stays last parameter?

Reply:
Heh, ok. I will revert the last commit and test different order.

Reply:
Unfortunately, moving the eager parameter did not work, just getting more errors. So I see 2 solutions:
1. Either wait until 2.12 is not supported by Akka anymore and merge this then.
2. Use only one apply method, but with default value `eagerCancel = false`, and ignore the MiMA warnings

I'm not even sure 2 can be considered by you, but having one apply with `eagerCancel = false` also makes it more consistent with other components, e.g. `Broadcast`

Reply:
Unfortunate indeed.

Default value is very problematic for binary compatibility so that we need to avoid, I think the options are to wait until we drop 2.12 support or add a factory method called something else than `apply` (for now).

Reply:
Yeah, I guess we wait for 2.12 then. I don't think a factory method is needed, would just confuse the API more. These changes were meant to make the `apply` the same as the regular class constructor. That `apply` hides arguments is not nice IMHO. But we can defo wait for 2.12 to be dropped.

I'll remove the last commit then, so that the PR is pure